File: algorithms/dynamic-programming/optimal-bst/swift/OptimalBST.swift
func optimalBst(_ arr: [Int]) -> Int {
    let n = arr[0]
    if n == 0 { return 0 }
    let freq = Array(arr[1...n])

    var cost = Array(repeating: Array(repeating: 0, count: n), count: n)
    for i in 0..<n { cost[i][i] = freq[i] }

    if n >= 2 {
        for len in 2...n {
            for i in 0...(n - len) {
                let j = i + len - 1
                cost[i][j] = Int.max
                var freqSum = 0
                for k in i...j { freqSum += freq[k] }

                for r in i...j {
                    let left = r > i ? cost[i][r - 1] : 0
                    let right = r < j ? cost[r + 1][j] : 0
                    let c = left + right + freqSum
                    if c < cost[i][j] { cost[i][j] = c }
                }
            }
        }
    }

    return cost[0][n - 1]
}

File: algorithms/dynamic-programming/optimal-bst/rust/optimal_bst.rs
pub fn optimal_bst(arr: &[i32]) -> i32 {
    let n = arr[0] as usize;
    let freq: Vec<i32> = arr[1..=n].to_vec();

    let mut cost = vec![vec![0i32; n]; n];
    for i in 0..n {
        cost[i][i] = freq[i];
    }

    for len in 2..=n {
        for i in 0..=(n - len) {
            let j = i + len - 1;
            cost[i][j] = i32::MAX;
            let freq_sum: i32 = freq[i..=j].iter().sum();

            for r in i..=j {
                let left = if r > i { cost[i][r - 1] } else { 0 };
                let right = if r < j { cost[r + 1][j] } else { 0 };
                let c = left + right + freq_sum;
                if c < cost[i][j] {
                    cost[i][j] = c;
                }
            }
        }
    }

    cost[0][n - 1]
}
